Softeq.XToolkit.PushNotifications.Droid/Services/DroidPushNotificationsConsumer.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Developed by Softeq Development Corporation
+// http://www.softeq.com
+
+using System.Threading.Tasks;
+using Android.App;
+using Android.Content;
+using AndroidX.Lifecycle;
+using Firebase.Messaging;
+using Softeq.XToolkit.Common.Extensions;
+using Softeq.XToolkit.Common.Logger;
+using Softeq.XToolkit.PushNotifications.Abstract;
+using Softeq.XToolkit.PushNotifications.Droid.Abstract;
+
+namespace Softeq.XToolkit.PushNotifications.Droid.Services
+{
+    /// <summary>
+    ///     Default implementation of <see cref="IPushNotificationsConsumer"/> interface for Android platform.
+    /// </summary>
+    public sealed class DroidPushNotificationsConsumer : IPushNotificationsConsumer
```

Comment:
Well, I agree that the library code should be extendable, but inheritance is not the only way. Current implementation uses DI, and basically all the work is done by dependencies, consumer just coordinates them. Replacing dependencies is the way of replacing implementation details and saving general logic and structure at the same time.
I can make the class non-sealed, but it won't be any difference without virtual members. Do you have in mind which methods of the consumer classes could be safely made virtual?

Reply:
`IosPushNotificationsConsumer.TryHandleRemoteNotification` seems to be a good choice. But that might be due to a mistake in previous implementation, actually. @wcoder Do you know why we didn't use this callback in the previous implementation, and used `UNUserNotificationCenterDelegate.WillPresentNotification(...)` to register received messages? It seems like it works with non-silent notification only

Reply:
And regarding `DroidPushNotificationsConsumer.TryHandleNotification`, it seems like we need to indeed implement a way to extend the way notifications are presented, but I don't thing virtual method is a good idea here (at least currently), because:
- `INotificationsSettingsProvider` is a required dependency off this class, all users should provide some implementation of this class;
- `INotificationsSettingsProvider` dependency is only used by `NotificationsHelper`, we don't even store a hard reference of it in the consumer;
- lets say, we implement a virtual `PresentNotification(PushNotificationModel)` method. If this method is ever overridden and new implementation does not use `NotificationsHelper` class - then `INotificationsSettingsProvider` won't be used either, it would become an unused required dependency, which is bad;

It seems like we actually need another Android-specific dependency, something like `INotiicationPresenter`, which might be used to override the way how notifications are presented. Also, this class might help with badge handling, it seems

Reply:
> I think it's about the behavior of the background/foreground:

Indeed, it seems like it's about background/foreground, not about silent/not silent. But still, current behavior seems to be the following:
- notifications are handled only when iOS app is in the foreground, if any notification is received in the background - no information will be passed to the app, unless notification is tapped or action is executed (but this is true only for not silent notifications, it seems, silent notifications in the background would be ignored completely);
- while in foreground, all received notification would be handled according to the `ForegroundNotificationOptions` value, even if notification is silent;
Is this the intended behavior?
https://github.com/Softeq/XToolkit.WhiteLabel/blob/4f4539ba0ebe05fca6e29acfd51eda70dba11b4d/Softeq.XToolkit.PushNotifications.iOS/UserNotificationCenterDelegate.cs#L23-L41

Reply:
I made some tests:
- message push (title, body):
  - background -> display notification, no methods calls
  - foreground -> displays notification, calls UNUserNotificationCenter delegate if declared otherwise `DidReceiveRemoteNotification` will be called.
- silent data push (custom data):
	- background -> only `DidReceiveRemoteNotification` call
	- foreground -> only `DidReceiveRemoteNotification` call

As result I see that new impl. will never call `handler.HandleSilentPushNotification`.

### `IosPushNotificationsConsumer.TryHandleRemoteNotification`
What about such implementation:
```cs
public bool TryHandleRemoteNotification(
            UIApplication application,
            NSDictionary userInfo,
            Action<UIBackgroundFetchResult> completionHandler)
{
    if (!_pushNotificationsParser.TryParse(userInfo, out var parsedNotification))
    {
        return false;
    }

    if (parsedNotification.IsSilent)
    {
        _pushNotificationsHandler.HandleSilentPushNotification(parsedNotification);
        return true;
    }
    else
    {
        if (UNUserNotificationCenter.Current.Delegate == null)
        {
            // need to call service constructor for init before
            // throw an exception or try to show a foreground notification if needed
        }
        else
        {
            // error - invalid behavior / log message
        }
    }

    return false;
}
```
or we can delegate this to the next consumer?

Reply:
I agree, we need to add notification handling into the `IosPushNotificationsConsumer.TryHandleRemoteNotification`, but these are my thoughts about implementation:

- regarding delegating to the next consumer. As I said, the general idea of the default consumers implementation is that successful parsing is the primary identification that notification belong to this particular consumer, thus it should not be passed to any other consumer;

- regarding `UNUserNotificationCenter.Current.Delegate` check - I understand now that this is a result of extracting `IPushNotificationAppDelegate` implementation into a separate, independent class. If this interface is implemented by the same `IosPushNotificationService` class it won't be possible that delegate is null - it would be initiated in the constructor of the class. And initialisation of the delegate, as I've mentioned before, is recommended to be done as soon as possible, ideally within `ApplicationFinishedLaunching()` method. So, I'd say we need to resolve two tasks here: (1) ensure `UNUserNotificationCenter.Current.Delegate` consistency (our implementation always uses this delegate, it should be always initialised while system is used, otherwise it might work incorrectly) and (2) enforce `UNUserNotificationCenter.Current.Delegate` to be initialised as soon as possible. Implementing two interfaces by single class would fix both, but maybe you have some other ideas?

- regarding how we detect if notification is silent. Currently we rely on parser to detect if notification is silent or not, but in reality there's only one "true" way of detecting it for each platform (there's a way of detecting it based on payload, but in some cases, like this one here, we know for sure that notification should be silent). This may cause some inconsistencies (again, like here, where we expect silent notification but theoretically can get both) and redundancy in API (for instance, `IPushNotificationsHandler.HandleSilentPushNotification(parsedNotification)` receives message that should have `IsSilent == true`, but it is not guaranteed by the signature of the method - so do implementors need to double check?) Both of these issues might be fixed by: (1) letting the consumer decide whether notification is silent or not based on context or payload (in this case it might be hardcoded because there's only one definitive way to do it on each platform) and pass notification to two different methods of the handler, or (2) keep parsing as is, but let the handler decide what to do with notification by itself - unite `HandlePushNotificationReceived` and `HandleSilentPushNotification` into a single method, within which it would be always easy to determine if notification silent or not. First solution would be more strict with the definition of "silent" push-notification, with the second it would be possible to use a custom definition of a "silent" push-notification.

What do you think about these points?

Reply:
I see that the push notifications structure looks like this:
- `IosPushNotificationService` - service for controlling push notifications in general as user;
- `IosPushNotificationAppDelegate` - just a default proxy;
- `IosPushNotificationsConsumer` - default consumer, which handles messages and states like a manager, if can and log warnings about strange messages or errors in default infrastructure.
- `IPushNotificationsHandler` - what I can do with messages/states as user.

regarding `UNUserNotificationCenter.Current.Delegate` check: I think that two separate classes it's good and explicit to understand and simple to customize. (2) point looks good to me in light mode and can be reached via warnings and documentation.

regarding silent notifications:
Yes, we have limitations with `HandleSilentPushNotification` because we have strict `PushNotificationModel` for that, I propose changing the method signature to use `object` as an argument.

Reply:
I strongly disagree with using the `object` in the interface. `IPushNotificationsHandler` is a common interface, declared in non-platform specific project, currently requires a single implementation - I can't imagine however it would be implemented with the `object` as an argument. How would it be possible to parse this `object` even? It won't be possible to cast it to any available class. Implement two handlers for each platform then? Seems like an overhead, and also `object` might be replaced with a platform-specific type, thus keeping our system strongly typed.

> (2) point looks good to me in light mode and can be reached via warnings and documentation.

If we know that a sertain state is invalid and should be avoided - isn't it better to just not let this state occure at all? Shouldn't we make it harder for users to make a wrong thing instead of just highlighting it?
Also, you insisted on splitting implementation into two classes to simplify the registration. But won't it be a complication to make user read through documentation and explain, why user needs to resolve `IPushNotificationService` before using `IPushNotificationAppDelegate`, even if it is not gonna be actually used?

Reply:
- `HandleSilentPushNotification`: ok, let's keep it as is, I have no cases for a better solution for now.
- `IPushNotificationService`,  `IPushNotificationAppDelegate`: The good case of trade-off, I prefer to keep these classes explicit for registration/use. And we should assume that each of them will be replaced. But from another side, we have sealed default implementations - and in the case single service class we should rewrite full service.

Reply:
Why should we assume that they would be replaced. They are implemented now in such a way, so they ensure correct work of consumers - all the callbaks are passed to the consumers, behavior can be changed/extended by implementing/combining consumers. Is it possible to implement those interfaces in any other way except passing calls to the consumers? No, it would brake the integrity of the system. Unless you want to implement an entirely new system - but the why would you need this package at all..
That is why I made those classes seald. The idea was to provide complete and robust base, so that actual behavior can be **safely** extended through consumers

Reply:
Ok, good point, let's keep a single service for all interfaces per ios/droid platform to be consistent. 
TODO: we need to expand Container for simply registering such cases.